### ISSUE TITLE
The call hierarchy UI should have the ability to filter calls, or apply name filter patterns

### ISSUE THREAD

**Issue by tommai78101** [microsoft/vscode#135972]
**Note**: This came from VS Code Java extension, referencing the issue here: https://github.com/redhat-developer/vscode-java/issues/2179

When I use **Show Call Hierarchy** on a property in a Java class object that is frequently used in both the main source code, as well as in the unit tests, the Call Hierarchy list would display a bunch of results. Because, the results include both the main source code and unit tests references, there isn't a way to automatically exclude certain folder directories or paths out from unwanted invocations.

I would like to request for a UI improvement in Visual Studio Code, to allow users apply any one of the following options:

- Filter test codes based on folder directory. (e.g., Filtering Java unit test codes or Java source code documents residing under `src/test/java` when specified)
- Apply name filter patterns. (e.g., Filtering reference names that matches the given filter patterns in the reference class names.)
- Limit the call depth of a given number. (e.g., Only goes down in depth up to the given number, inclusive.)

This feature is similar to Eclipse IDE's "Filter Calls" dialog:

![image](https://user-images.githubusercontent.com/6444116/138523340-7b14bfe2-5463-4fa7-adf3-92118ecde522.png)

This is accessed by invoking "Show Call Hierarchy" on a class member, then opening "Filters" in the "More Options" menu in Eclipse IDE.

Currently, we don't have anything similar to this in Visual Studio Code in general (regardless of what language support), where the call hierarchy UI allows the user to filter out what references to look for / ignore.

##### Environment
- Operating System: Ubuntu 20.04.2 LTS
- JDK version: AdoptOpenJDK 11  (jdk-11.0.11+9)
- Visual Studio Code version:  1.61.2

##### Current Result

The "Show Call Hierarchy" looks something like this, whenever we attempt to invoke call hierarchy. I'm using Java language support as an example.

![image](https://user-images.githubusercontent.com/6444116/138524191-a7c98040-2fdb-4ad3-aa27-11a118db7a4b.png)

##### Expected Result

Call hierarchy UI with filter options would automatically filter out either all test codes, or applying a name filter pattern to exclude any references containing "test". A mock example should look something like this immediately upon invoking the call hierarchy:

![image](https://user-images.githubusercontent.com/6444116/138524377-a1d0f202-1fef-4e42-a683-9dffc20c3104.png)

**Comment by cortexcompiler** [comment]
This would be very helpful.  IntelliJ has a similar feature allowing you to filter to just "production" code.  The include and exclude capability that is already within vscode Search would do the trick:
![image](https://user-images.githubusercontent.com/12128840/143064870-861d56ab-cd68-463a-8692-6b109d258fd9.png)

**Comment by OWissett** [comment]
It would also be nice to be able to filter based on a regex for function signatures

**Comment by NeonWizard** [comment]
Bump on this. Is this being tracked in a newer issue or is it just stale? This feature seems to be widely requested on stack overflow and other help sites. @jrieken

**Comment by tommai78101** [comment]
@NeonWizard If you could help give this feature at the top with a thumbs up, it will allow them to prioritize this feature better. The more thumbs up, the more awareness this feature request will get, the more prioritization.

**Comment by j1elo** [comment]
What surprises me is that apparently only 13 people (current count of upvotes) use the Show Hierarchy feature? Because the minute I discovered and used it for the first time today with my Go project, I thought that it'd be great to filter tests out. Hence here I am, after searching to find if someone had already made the suggestion.

Given that it was immediately apparent to me, as first time user, I definitely expected that it would be for other people too, and this was a more popular feature request. Well I guess not :)

**Comment by tommai78101** [comment]
@j1elo I'm pretty sure a lot of people uses the Call Hierarchy feature. They probably just don't realize that filtering the method invocations and classes would help them more.

**Comment by dreyjey1993** [comment]
We have very long lists of calls and about 90% of the results are log calls, which make it difficult to find other calls.

So it would be good if there was also the option to exclude results.
I would go one step further and implement sorting in addition to filtering.

There could be the following sorting options:
* By call sequence (default)
* By name
* By type/file extension

When sorting by name or type, the results found would be grouped.

**Comment by callain-mirakl** [comment]
I was looking for this feature today, it makes debugging harder when it shows all the tests in the hierarchy

**Comment by NeonWizard** [comment]
@VSCodeTriageBot We need this issue to be reassigned, Johannes seems to have gone AWOL

**Comment by tommai78101** [comment]
@NeonWizard Johannes isn't technically AWOL. He's still active as a of last week. It's just that this issue is staying in the backlog, and it hasn't been prioritized as much as the rest of the features that are being put into the upcoming monthly milestone goals.

**Comment by catskul** [comment]
Just mentioning closed version of this so there's a breadcrumb for others who found the other dead issue: https://github.com/microsoft/vscode/issues/205411

**Comment by fisherbe** [comment]
This would be great for Python to exclude all third party packages and builtins for sure!

**Comment by aadil42** [comment]
Yes! I was looking for this feature is it merged yet? I don't see it in my vs code. 

<img width="385" height="676" alt="Image" src="https://github.com/user-attachments/assets/75e0ba7b-af25-4b93-bf87-5791cb5a2558" />
I was looking for specific callers from the list of callers as you can see it is a long list with nested calls. The feature to filter out based on fileName, folderName, callerFunctionName would be of huge help.

**Comment by aadil42** [comment]
@jrieken, Let me know if I can be of any help. I have some time and I would love to help implement this feature in any way. This feature would be very helpful. 
Thank you.
